### ISSUE TITLE
v3: Support Programmatically Rewriting Ledger

### ISSUE THREAD

**Issue by aclindsa** [beancount/beancount#586]
As discussed in https://groups.google.com/g/beancount/c/w-LBWdSbmLw/m/JqKn8dR-AwAJ, it would be nice to provide a mechanism to allow users to programmatically re-write their ledgers. A few potential use cases I'm thinking of:

1. Re-ordering imported transactions by date
2. Re-formatting a hand-edited ledger (possibly replacing `bean-format`?)
3. Splitting transactions into different files by type (i.e. ledger.beancount, prices.beancount, and balances.beancount) or associated accounts (i.e. maintaining credit_card.beancount for all credit card transactions)
4. More sophisticated transformations (find all transactions with a posting against `Income:PnL`, and change them to be of the form `Income:PnL:Personal-Brokerage`, where `Personal-Brokerage` corresponds to the other postings in the transaction)
5. Cleaning up old data (maybe you realized that many transactions you imported a while ago from GnuCash have metadata which partially duplicates their narration and you want to remove it if so)

@blais 's initial suggestion on the mailing list was to 
> provide some library function to accept a file and produce a mapping of
> 
>   (parser-directive, text)
>
> whereby the parser directive is the unprocessed, straightforward translation of your input text to the intermediate data structure

My current understanding is that there are likely 2-3 primary changes to allow this:

1. Update the parser to parse comments into a simple data structure representing them
2. Adding a library function which would expose the intermediate representation of all parsed directives (including comments and config options) given an input file
3. A printer/formatter for the exposed intermediate data structures

Note: The third (printer for intermediate data structures) is perhaps a stretch goal of mine beyond what was previously discussed, but I believe it would be needed to allow some of the more sophisticated transformations I mentioned above.

**Comment by zacchiro** [comment]
A couple more use cases from my personal experience (which I had to address with horrible sed/perl hacks…):
1) renaming accounts/restructuring the account hierarchy, without capture and without messing up indentation
2) normalizing the order in which metadata appear in a transaction or in legs of it
3) moving specific metadata "up" from transaction legs to the transaction as a whole

**Comment by blais** [comment]
Yes I think I'll implement this one, this will allow a whole new level of applications that process the raw input. 

It'll require a few more changes, like not evaluating the arithmetic expressions early (in order to be able to write them back out), but this also will provide an opportunity to set the decimal precision context, so I've had it in the back of my mind anyway.

**Comment by polarmutex** [comment]
I am working on a Language Protocol Server(LSP) for Beancount. I have a lot of these on my todo list. The LSP is based on tree-sitter which can handle these tasks pretty easy. I can post back on this issue when I get some of these implemented

**Comment by hitit-web** [comment]
I would love this as well. In starting my beancount journey, I realized very early that I like to have my files more orderly than I am capable of producing by hand plus to have something that auto-sorts transactions by date.

To that end, I have been avoiding comments of any kind and on linux I am currently passing my beancount files through `bean-report <file>.beancount print > temp.beancount` and then replacing my the old file with the temp file. This makes very nice formatting but is a bit of a pain because I have all of my account opening statements in a separate file. I have to concat that file to the account file, run the report and separate it out again.

I would very much appreciate an expanded "bean-format" to take care of the use cases mentioned in previous posts.

**Comment by umangv** [comment]
I'm wondering if [Martin's message on the mailing list](https://groups.google.com/g/beancount/c/SCZ2VSTDLQI/m/pKzWxj_cBwAJ) is the latest status on this. I know [autobean-refactor](https://autobean-refactor.readthedocs.io/en/latest/getting-started.html) exists. Does it make sense to make that the "official" implementation of this functionality (mostly to know that if I write an importer that relies on that, it would work well with whatever comes in the future)? I'm wondering if there are any gaps between the aspirations for this feature and what exists today with autobean-refactor?

**Comment by blais** [comment]
The latest is that the C++ branch has a much better parser and data
structure (separate for parsed syntax with missing parts vs. the final
result which should have been interpolated and booked, as it should be),
and extending that to include and preserve comments would be a possible
incremental change there (though not entirely trivial). I think one wants
separate data structures for uninterpolated incomplete input over the
finalized, completed, booked, validated stream of transactions. However,
I've put the C++ port on hold, due to limitations with passing protobufs
between Python and C++ and not enough time to work on this.

I'm more likely to eventually create a Rust version, at which point I would
want to have a new parser that includes comments and can round-trip the
entire file byte-for-byte. That would be a neat way to make changes
programmatically: run the parser, modify the data structure, and write it
back.  (Of course the parser would not include the effect of plugins,
interpolation, booking, etc, just the information provided in the file.)

There are a few parsers in Rust for Beancount syntax. I remember testing
one out a while ago and working with the author via tickets to identify
shortcomings and limitations and I believe it covers all the syntax in my
large ledger.  It might be a fun idea to extend that parser to preserve
comments maybe?
